Update documentation private fork

Lars-DHondt-KUL opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Provide a short overview of the changes -->
Explain how to:
1. update the master branch of  a private repo based on the public repo
2. add code from a private repo to the public repo with or without write access to the public repo

Create a batch script to perform 1., since the steps are always exactly the same.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This info was living in a file on my laptop.

## How Has This Been Tested?
<!--- Please describe how you tested your changes. -->
<!--- Please describe the outcome of these tests. -->
All previous movements between both repos were according to these steps.

## Suggested tests for reviewers
<!---Please describe tests for the reviewer to run -->
- proofread
- is the explanation clear? (as far as these github shenanigans can be clear)
- update the master branch of PredSim_private (after public has been changed) Try automated way first.
- bring a (dummy) branch from the private to the public repo

<!--- Assign labels ("Labels" tab in side bar). Each PR should have at least one "Review:..." label, since we use these to allocate reviewers. -->
<!--- Do not add requested reviewers, unless this person specifically agreed to review your PR. -->
<!--- If you are a collaborator (i.e. have direct write access to this repo) select yourself as assignee, otherwise someone will be assigned to your PR. -->

Lars-DHondt-KUL committed:
Create update_from_public.bat

Lars-DHondt-KUL committed:
Explain how to copy code between private and public repos

Lars-DHondt-KUL committed:
improve formatting

VDB-Bram reviewed (commented):
First general feedback:
1. good that this information is added
2. it became only clear later that the numbered steps in the figure align with the numbered steps in the text; should this be further specified? (or it is just me)
3. comments in the `.bat` file could help users understand what is happening

VDB-Bram commented:
Testing the manual workflow to get changes to the private.

1. a branch called `master_public` should already exist; otherwise the commands are: 
`$ git checkout master`
`$ git checkout -b master_public` this creates the new branch
`$ git push -u origin master_public` to make sure the branch in published, otherwise step 4 (`$ git push origin`) does not work.
2. in step 3 I would make the commit message `"update from public master"` (although I do not see the commit message anywhere)

But in general, the manual workflow seems to work.

VDB-Bram commented:
Test bringing a (dummy) branch from the private to the public repo.

Straight from the private to the public using `$ git push upstream amazing_new_code`: succesful, see https://github.com/KULeuvenNeuromechanics/PredSim/tree/dummy-branch

From private to an existing branch in public using `$ git push upstream tmp_3-final2:amazing_new_code`: unsuccesful. It seems though that this work: `$ git push upstream private_banch_name:public_branch_name`.  See https://github.com/KULeuvenNeuromechanics/PredSim/tree/dummy-test2
I think that this option should not be encouraged though. People might change branches where other people are working on.
 
_both dummy branches can be deleted after the approval of this PR_

Lars-DHondt-KUL committed:
Update PrivateForkPredSim.md

Lars-DHondt-KUL committed:
Update update_from_public.bat

VDB-Bram commented on Documentation/PrivateForkPredSim.md:
```diff
@@ -34,8 +36,8 @@ For this repository the Git Bash commands are:
-   
- 6. If you want, add the original repo as remote to fetch (potential) future changes.
+
+ 6. If you want, add the original repo as remote to fetch (potential) future changes. This creates the diagonal link in the figure below.
```
Since there are also arrows between the local and remote, maybe have different color or something such that you can clearly indicate where the digital link is?

VDB-Bram commented on VariousFunctions/update_from_public.bat:
```diff
@@ -0,0 +1,101 @@
+	echo git checkout -B master_public
+	echo git push -u origin master_public
+	rem % 3.
+	echo git merge %PUBLIC_REPO_REMOTE_NAME%/master -m "update from public master"
```
I have been testing some things; in the case that `master_public` is ahead of `upstream/master`, the branch will not be reset to be the exact same as `upstream/master`. Therefore, `master` in the private repo will not be the same as `master` in the public repo. 

This is something I noticed when testing how I could make sure that `master_public` was exactly the same as `master` in the private repo before merging `upstream/master` into it. I tried `$ git checkout -B test_branching master` but this did not seem to work.

Lars-DHondt-KUL replied:
> After `$ git checkout -B master_public master` I ran `$ git push origin master_public` but this resulted in an error.

What if you run `$ git push -u origin master_public` instead?
I'm not sure [what exactly it does](https://git-scm.com/docs/git-push#Documentation/git-push.txt--u).

If that doesn't help:

> This issue was resolved when running `$ git push origin master_public --force` instead.

Or should we use `$ git push origin +master_public`?
This might be safer, as it [only forces a push to the `master_public` branch](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

Lars-DHondt-KUL committed:
Merge branch 'master' into update-documentation-private-fork

Lars-DHondt-KUL committed:
Update PrivateForkPredSim.md

Added figure captions

Lars-DHondt-KUL committed:
Update PrivateForkPredSim.md

add clarification
see https://github.com/KULeuvenNeuromechanics/PredSim/pull/203#issuecomment-3180077270

Lars-DHondt-KUL committed:
Update PrivateForkPredSim.md

Lars-DHondt-KUL committed:
clean up update_from_public.bat

ellisvancan commented on Documentation/PrivateForkPredSim.md:
```diff
@@ -1,6 +1,8 @@
+
 Workflow below is adapted from [here](https://gist.github.com/0xjac/85097472043b697ab57ba1b1c7530274).
 
 The correct way of creating a private frok by duplicating the repo is documented [here](https://help.github.com/articles/duplicating-a-repository/).
```
typo - frok instead of fork

ellisvancan commented:
Notes Francesco: 
# Creating the private repository
4. Remove the temporary local repository you created in step 1
MY NOTE: The code includes a "rm" command, which is a GitBash command. However, it was not clear to me that this code should have been run in a GitBash instead of a Windows command prompt console. In other parts of the page,  whenever a GitBash console is specifically needed, this is clearly stated.

# Updating PredSim_private based on the public PredSim repository
## Automated
MY NOTE: in the cloned version, the update_from_public.bat file is missing

## Manual
2. In the private repository on your computer, make master_public the active branch.
MY NOTE: it is not clear what master_public is here git gives error " did not match any file(s) known to git"

Also
$ git checkout -B master_public master
gives
fatal: 'master' is not a commit and a branch 'master_public' cannot be created from it

Lars-DHondt-KUL committed:
Update PrivateForkPredSim.md

Lars-DHondt-KUL committed:
Merge branch 'master' into update-documentation-private-fork

Lars-DHondt-KUL committed:
Update PrivateForkPredSim.md

Lars-DHondt-KUL commented:
> # Creating the private repository
> 
> 4. Remove the temporary local repository you created in step 1
> MY NOTE: The code includes a "rm" command, which is a GitBash command. However, it was not clear to me that this code should have been run in a GitBash instead of a Windows command prompt console. In other parts of the page,  whenever a GitBash console is specifically needed, this is clearly stated.

I added a *step 0* to make this more explicit

> # Updating PredSim_private based on the public PredSim repository
> ## Automated
> 
> MY NOTE: in the cloned version, the update_from_public.bat file is missing


Because your private fork is based on the *master* branch of PredSim. After this pull request is done, the new file will be added in *master*, so new private forks will have the file. For private forks that were made before that (like the test now), you will need to copy the file manually

> ## Manual
>  2. In the private repository on your computer, make master_public the active branch.
>        MY NOTE: it is not clear what master_public is here git gives error " did not match any file(s) known to git"

I updated the description of this step.

> Also $ git checkout -B master_public master gives fatal: 'master' is not a commit and a branch 'master_public' cannot be created from it

Before doing `git checkout -B master_public master`,
try `git fetch origin`
If that does not fix it, try to also run `git checkout master` or `git checkout origin/master`

If any of these extra commands result in successfully running `git checkout -B master_public master` afterwards, I will add them to the description.

InesVandekerckhove reviewed (approved):
instructions are clear to me
